### PR TITLE
chore(cli): validate various names locally

### LINKED ISSUE #10244: Improve `Validation failed` error message for template names with underscores
In most programming languages I'm used to, a lower dash '_' is handled like a character and a normal dash '-' ist not allowed in variable names.  With the template names it's the other way around: a '-' is allowed and a '_' leads to a cryptic "Validation failed" after terraform ran. 

I would suggest to raise some kind of warning, when the template name is invalid.

### LINKED ISSUE #14544: CLI: `coder create` does not check/sanitize workspace name before creating
[Discord thread](https://discord.com/channels/747933592273027093/1252546885466587237)

---

`coder create` checks for the name after running the `terraform plan`, which is fine in most cases even though it makes you lose some time.

Here is what would usually happen to an user :
<details>
  <summary>output of <code>coder create phorcys420/invalid_name -t aws-linux</code> (on dogfood)</summary>

  ```
  PS C:\Users\Phorcys\Documents> coder create phorcys420/invalid_name -t aws-linux
  AWS Region
    The region to deploy workspace infrastructure.
  ?  EU (Paris)
  
  > EU (Paris)
  Instance type
    What instance type should your workspace use?
  ?  2 vCPU, 1 GiB RAM
  
  > 2 vCPU, 1 GiB RAM
  Planning workspace...
  === ✔ Queued [0ms]
  ==> ⧗ Running
  ==> ⧗ Running
  === ✔ Running [13ms]
  ==> ⧗ Setting up
  === ✔ Setting up [69ms]
  ==> ⧗ Detecting persistent resources
  === ✔ Detecting persistent resources [15661ms]
  ==> ⧗ Cleaning Up
  === ✔ Cleaning Up [39ms]
  ┌──────────────────────────────────────────────────────────┐
  │ Workspace Preview                                        │
  ├──────────────────────────────────────────────────────────┤
  │ RESOURCE                    ACCESS                       │
  ├──────────────────────────────────────────────────────────┤
  │ aws_ec2_instance_state.dev                               │
  ├──────────────────────────────────────────────────────────┤
  │ aws_instance.dev                                         │
  │ └─ dev (linux, amd64)        coder ssh invalid_name      │
  └──────────────────────────────────────────────────────────┘
  > Confirm create? (yes/no) yes
  Encountered an error running "coder create", see "coder create --help" for more information
  error: Trace=[create workspace: ]
  Validation failed.
  1 validation error(s) found
           name : Validation failed for tag "workspace_name" with value: "invalid_name"
  ```
</details>

But in some cases, it will even make the plan fail if the workspace name causes an error in the template, such as with `_` in k8s.
Which makes it so that it doesn't even show the validation error because the plan exits before the validation.

<details>
  <summary>output of <code>coder create</code> commands with the <a href="https://github.com/coder/coder/tree/main/examples/templates/kubernetes">sample Kubernetes template</a> (on my testing instance)</summary>

```
PS C:\Users\Phorcys> coder create admin/workshop_ia -t kubernetes-titimoby
version mismatch: client v2.11.2+d53c94b, server v2.10.2+a11b169
download the server version from: https://github.com/coder/coder/releases/v2.10.2
Planning workspace...
=== ✔ Queued [70ms]
==> ⧗ Running
==> ⧗ Running
=== ✔ Running [36ms]
==> ⧗ Setting up
=== ✔ Setting up [114ms]
==> ⧗ Detecting persistent resources
=== ✔ Detecting persistent resources [6134ms]
==> ⧗ Cleaning Up
=== ✘ Cleaning Up [12ms]
=== ✘ Cleaning Up [31ms]
Encountered an error running "coder create", see "coder create --help" for more information
error: prepare build: dry-run workspace: run dry-run provision job: terraform plan: exit status 1

PS C:\Users\Phorcys> coder create admin/workshop-ia -t kubernetes-titimoby
version mismatch: client v2.11.2+d53c94b, server v2.10.2+a11b169
download the server version from: https://github.com/coder/coder/releases/v2.10.2
Planning workspace...
=== ✔ Queued [64ms]
==> ⧗ Running
==> ⧗ Running
=== ✔ Running [12ms]
==> ⧗ Setting up
=== ✔ Setting up [106ms]
==> ⧗ Detecting persistent resources
=== ✔ Detecting persistent resources [7152ms]
==> ⧗ Cleaning Up
=== ✔ Cleaning Up [93ms]
┌──────────────────────────────────────────────────────────────────────┐
│ Workspace Preview                                                    │
├──────────────────────────────────────────────────────────────────────┤
│ RESOURCE                                 ACCESS                      │
├──────────────────────────────────────────────────────────────────────┤
│ kubernetes_deployment.main                                           │
│ └─ main (linux, amd64)                    coder ssh workshop-ia      │
├──────────────────────────────────────────────────────────────────────┤
│ kubernetes_persistent_volume_claim.home                              │
└──────────────────────────────────────────────────────────────────────┘
```

---

This should be a relatively low-effort change in my opinion.

### PR DESCRIPTION
**Pull request by ethanndickson**
Closes #14544.
Closes #10244.

As mentioned in https://github.com/coder/terraform-provider-coderd/pull/85, `coderd` doesn't currently provide an error reason when a field in a request [fails to validate](https://github.com/go-playground/validator/issues/1093). This means the CLI provides an arguably useless error message when a supplied name is invalid. 

Even if `coderd` supplied an error reason when a field failed to validate, the problem in #14544 would still result in a not great UX, only showing the error after a plan (which could reasonably take several minutes for some Terraform providers). 

I think it's okay to follow the Web UI here, and validate locally first, especially since both `coderd` and the CLI will call the same validator implementation in `codersdk`.

### CODE REVIEW

**Comment by ethanndickson** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/14551?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14551** <a href="https://app.graphite.dev/github/pr/coder/coder/14551?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14550** <a href="https://app.graphite.dev/github/pr/coder/coder/14550?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ethanndickson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: cli/create.go [hunk 1]
```diff
@@ -60,7 +60,11 @@ func (r *RootCmd) create() *serpent.Command {
 				workspaceName, err = cliui.Prompt(inv, cliui.PromptOptions{
 					Text: "Specify a name for your workspace:",
 					Validate: func(workspaceName string) error {
-						_, err = client.WorkspaceByOwnerAndName(inv.Context(), codersdk.Me, workspaceName, codersdk.WorkspaceOptions{})
+						err = codersdk.NameValid(workspaceName)
+						if err != nil {
+							return xerrors.Errorf("Workspace name %s", err)
```
Comment (deansheather) [inline cli/create.go:65]: Every where you do these errors you should do something like: `"workspace name %q is invalid: %w"`

Error wraps should also always be used with `%w` and be at the end with a colon in front, like `"message: %w"`, and error messages should not start with a capital letter

File: enterprise/cli/groupcreate.go [hunk 2]
```diff
@@ -35,6 +35,11 @@ func (r *RootCmd) groupCreate() *serpent.Command {
 				return xerrors.Errorf("current organization: %w", err)
 			}
 
+			err = codersdk.GroupNameValid(inv.Args[0])
+			if err != nil {
+				return xerrors.Errorf("group name: %w", err)
```
Comment (deansheather) [inline enterprise/cli/groupcreate.go:40]:
```suggestion
				return xerrors.Errorf("group name %q is invalid: %w", inv.Args[0], err)
```

**Review by deansheather — APPROVED** [review]

File: enterprise/cli/groupcreate.go (hunk 2 above)
Comment (ethanndickson) [reply enterprise/cli/groupcreate.go:40]: So upset I missed one 😭

**Pull request merged**